ath79: add support for NEC Aterm WG1200CR

musashino205 opened the pull request:
NEC Aterm WG1200CR is a 2.4/5 GHz band 11ac router, based on Qualcomm
Atheros QCA9563.

Specification:

- SoC		: Qualcomm Atheros QCA9563
- RAM		: DDR2 128 MiB
- Flash		: SPI-NOR 8 MiB
- WLAN		: 2.4/5 GHz 2T2R
  - 2.4 GHz: QCA9563 (SoC)
  - 5 GHz  : QCA9888
- Ethernet	: 2x 10/100/1000 Mbps
  - Switch : QCA8334
- LEDs/keys	: 12x/4x (2x buttons, 1x slide-switch)
- UART		: through-hole on PCB
  - JP1: Vcc, GND, NC, TX, RX from power connector side
  - 115200 bps

Flash instruction using factory image:

1. Boot WG1200CR normaly
2. Access to "http://192.168.10.1/" and open firmware update page
("ファームウェア更新")
3. Select the OpenWrt factory image and click update ("更新") button
to perform firmware update
4. Wait ~150 seconds to complete flashing

Known issues:

- cannot be controlled LEDs other than Power (Green/Red)
  - only Power LEDs are connected to SoC GPIO; other LEDs connected to
    the gpiochip on ath10k chip (QCA9888)

Note:

- The U-Boot on WG1200CR loads Linux Kernel as a uImage
  - log:
```
## Booting image at 9f080040 ...
   Image Name:   MIPS Seattle Linux-3.3.8
   Created:      2018-11-05   9:39:44 UTC
   Image Type:   MIPS Linux Multi-File Image (lzma compressed)
   Data Size:    1064192 Bytes =  1 MB
   Load Address: 80060000
   Entry Point:  80060000
   Contents:
   Image 0:  1064184 Bytes =  1 MB
   Verifying Checksum at 0x9f080080 ...OK
   Uncompressing Multi-File Image ... OK
No initrd
## Transferring control to Linux (at address 80060000) ...
## Giving linux memsize in bytes, 134217728
 
Starting kernel ...
```

  - dumped OEM firmware:
```
$ hexdump -n 128 -C mtd5_firmware.bin
00000000  5e a3 a4 17 00 00 00 24  00 71 70 20 38 08 72 93  |^......$.qp 8.r.|
00000010  ed 3f fa 42 bf 9a e5 83  08 6e 93 d5 64 65 76 3d  |.?.B.....n..dev=|
00000020  2f 64 65 76 2f 6d 74 64  62 6c 6f 63 6b 2f 36 00  |/dev/mtdblock/6.|
00000030  74 79 70 65 3d 66 69 72  6d 77 61 72 65 00 00 00  |type=firmware...|
00000040  27 05 19 56 eb ac 39 da  5b e0 0f e0 00 10 3d 00  |'..V..9.[.....=.|
00000050  80 06 00 00 80 06 00 00  1e f7 e4 68 05 05 04 03  |...........h....|
00000060  4d 49 50 53 20 53 65 61  74 74 6c 65 20 4c 69 6e  |MIPS Seattle Lin|
00000070  75 78 2d 33 2e 33 2e 38  00 00 00 00 00 00 00 00  |ux-3.3.8........|
00000080
```

musashino205 committed:
firmware-utils: add nec-enc

nec-enc provides firmware encoding/decoding with model specific key
for NEC devices.

known devices:

  - Aterm WF1200CR
  - Aterm WG1200CR
  - Aterm WG2600HS

usage:

  nec-enc -i infile -o outfile -k key

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
ath79: add support for NEC Aterm WG1200CR

NEC Aterm WG1200CR is a 2.4/5 GHz band 11ac router, based on Qualcomm
Atheros QCA9563.

Specification:

- SoC		: Qualcomm Atheros QCA9563
- RAM		: DDR2 128 MiB
- Flash		: SPI-NOR 8 MiB
- WLAN		: 2.4/5 GHz 2T2R
  - 2.4 GHz: QCA9563 (SoC)
  - 5 GHz  : QCA9888
- Ethernet	: 2x 10/100/1000 Mbps
  - Switch : QCA8334
- LEDs/keys	: 12x/4x (2x buttons, 1x slide-switch)
- UART		: through-hole on PCB
  - JP1: Vcc, GND, NC, TX, RX from power connector side
  - 115200 bps

Flash instruction using factory image:

1. Boot WG1200CR normaly
2. Access to "http://192.168.10.1/" and open firmware update page
("ファームウェア更新")
3. Select the OpenWrt factory image and click update ("更新") button
to perform firmware update
4. Wait ~150 seconds to complete flashing

Known issues:

- cannot be controlled LEDs other than Power (Green/Red)
  - only Power LEDs are connected to SoC GPIO; other LEDs connected to
    the gpiochip on ath10k chip (QCA9888)

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

chunkeey commented:
```
nec-enc.c: In function ‘main’:
nec-enc.c:111:23: warning: pointer targets in passing argument 1 of ‘xor_pattern’ differ in signedness [-Wpointer-sign]
   k_off = xor_pattern(buf_pattern, n, key, k_len, k_off);
                       ^~~~~~~~~~~
nec-enc.c:18:5: note: expected ‘uint8_t *’ {aka ‘unsigned char *’} but argument is of type ‘char *’
 int xor_pattern(uint8_t *data, size_t len, const uint8_t *key, int k_len, int k_off)
     ^~~~~~~~~~~
nec-enc.c:111:39: warning: pointer targets in passing argument 3 of ‘xor_pattern’ differ in signedness [-Wpointer-sign]
   k_off = xor_pattern(buf_pattern, n, key, k_len, k_off);
                                       ^~~
nec-enc.c:18:5: note: expected ‘const uint8_t *’ {aka ‘const unsigned char *’} but argument is of type ‘const char *’
 int xor_pattern(uint8_t *data, size_t len, const uint8_t *key, int k_len, int k_off)
     ^~~~~~~~~~~
nec-enc.c:113:12: warning: pointer targets in passing argument 1 of ‘xor_data’ differ in signedness [-Wpointer-sign]
   xor_data(buf, n, buf_pattern);
            ^~~
nec-enc.c:29:6: note: expected ‘uint8_t *’ {aka ‘unsigned char *’} but argument is of type ‘char *’
 void xor_data(uint8_t *data, size_t len, const uint8_t *pattern)
      ^~~~~~~~
nec-enc.c:113:20: warning: pointer targets in passing argument 3 of ‘xor_data’ differ in signedness [-Wpointer-sign]
   xor_data(buf, n, buf_pattern);
                    ^~~~~~~~~~~
nec-enc.c:29:6: note: expected ‘const uint8_t *’ {aka ‘const unsigned char *’} but argument is of type ‘char *’
 void xor_data(uint8_t *data, size_t len, const uint8_t *pattern)
      ^~~~~~~~
```

musashino205 commented:
Thank you for the review, I'll fix it.

musashino205 commented:
Done, also fixed following warning:

```
src/nec-enc.c: In function 'main':
src/nec-enc.c:89:10: warning: 'key' may be used uninitialized in this function [-Wmaybe-uninitialized]
  k_len = strlen(key);
          ^~~~~~~~~~~
```

chunkeey commented:
checkpatch.pl also found some issues.

```
WARNING: Missing or malformed SPDX-License-Identifier tag in line 1
#44: FILE: tools/firmware-utils/src/nec-enc.c:1:
+/*

WARNING: line over 80 characters
#61: FILE: tools/firmware-utils/src/nec-enc.c:18:
+int xor_pattern(uint8_t *data, size_t len, const char *key, int k_len, int k_off)

ERROR: switch and case should be at the same indent
#101: FILE: tools/firmware-utils/src/nec-enc.c:58:
+               switch (c) {
+                       case 'i':
[...]
+                       case 'o':
[...]
+                       case 'k':
[...]
+                       case 'h':
+                       default:

ERROR: do not use assignment in if condition
#122: FILE: tools/firmware-utils/src/nec-enc.c:79:
+       if (ifn && !(in = fopen(ifn, "r"))) {

ERROR: do not use assignment in if condition
#127: FILE: tools/firmware-utils/src/nec-enc.c:84:
+       if (ofn && !(out = fopen(ofn, "w"))) {

WARNING: space prohibited between function name and open parenthesis '('
#139: FILE: tools/firmware-utils/src/nec-enc.c:96:
+               fprintf (stderr, "key length is incorrect\n");

WARNING: labels should not be indented
#146: FILE: tools/firmware-utils/src/nec-enc.c:103:
+                       FREAD_ERROR:

ERROR: trailing statements should be on next line
#155: FILE: tools/firmware-utils/src/nec-enc.c:112:
+                       if (ptn > 250) ptn = 0;

WARNING: labels should not be indented
#163: FILE: tools/firmware-utils/src/nec-enc.c:120:
+               FWRITE_ERROR:

WARNING: braces {} are not necessary for single statement blocks
#169: FILE: tools/firmware-utils/src/nec-enc.c:126:
+       if (ferror(in)) {
+               goto FREAD_ERROR;
+       }

WARNING: braces {} are not necessary for single statement blocks
#173: FILE: tools/firmware-utils/src/nec-enc.c:130:
+       if (fflush(out)) {
+               goto FWRITE_ERROR;
+       }

total: 4 errors, 6 warnings, 413 lines checked
```

But before you jump to action I just fixed them locally 

https://git.openwrt.org/?p=openwrt/staging/chunkeey.git;a=commit;h=da268215dd2034d075837801a3e882934c1dcf31

chunkeey commented:
> - cannot be controlled LEDs other than Power (Green/Red)
  - only Power LEDs are connected to SoC GPIO; other LEDs connected to
    the gpiochip on ath10k chip (QCA9888)

Dang. This used to work with the ath10k gpio patches. But as of late there's no support
for gpio access through ath10k-ct driver anymore since it was never ported to the later
releases.

chunkeey commented on target/linux/ath79/dts/qca9563_nec_wg1200cr.dts:
```diff
@@ -0,0 +1,176 @@
+			debounce-interval = <60>;
+		};
+
+		br {
```
Is this the "bridge" setting? If the device has a "br" printed on the box please update the node name to "bridge" and leave the label as "br".

musashino205 replied:
Yes, it means "bridge" and printed as "BR" on the box. OK, I'll change it.

chunkeey commented on target/linux/ath79/dts/qca9563_nec_wg1200cr.dts:
```diff
@@ -0,0 +1,176 @@
+			debounce-interval = <60>;
+		};
+
+		cnv {
```
converter?

musashino205 replied:
Yes, I'll change.

chunkeey commented on target/linux/ath79/dts/qca9563_nec_wg1200cr.dts:
```diff
@@ -0,0 +1,176 @@
+	status = "okay";
+
+	wifi@0,0 {
+		compatible = "pci168c,0056";
```
FYI: (you don't need to change anything)

https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/net/wireless/qcom%2Cath10k.txt
specifies that this should be "qcom,ath10k". But again "pci168c,0056" is fine too.

musashino205 replied:
OK, I'll keep in mind.

musashino205 commented:
I updated node names.

> checkpatch.pl also found some issues.
> But before you jump to action I just fixed them locally
Thank you!

> Dang. This used to work with the ath10k gpio patches. But as of late there's no support
> for gpio access through ath10k-ct driver anymore since it was never ported to the later
> releases.
Oh, I see...

chunkeey commented:
Merged. Thanks.

adrianschmutzler commented:
Although this PR has been has been closed already, I would be interested in some additional data about the device(s) in case you still have access to it:

I'm currently working on a PR to store the label-mac-address in the DTS (https://github.com/openwrt/openwrt/pull/2159).

I would appreciate if you could provide information about your device:
Please run
`find /proc/device-tree/ -name "*mac-address*"`
on your device (just with the current firmware, no need to update anything; just has to be ath79, no ar71xx).
For each of the returned paths (if there are any), retrieve the mac-address with `cat`, e.g.
`cat /proc/device-tree/ahb/eth@19000000/mac-address`

Now check which of those returned MAC addresses matches the one on the label/sticker/case of your device and tell me the device-tree path leading to it. Note that you do not have to post your MAC addresses here, but just a statement about which match, e.g. like
```
/proc/device-tree/ahb/eth@1a000000/mac-address
/proc/device-tree/ahb/eth@1a000000/mtd-mac-address-increment
/proc/device-tree/ahb/eth@1a000000/mtd-mac-address
/proc/device-tree/ahb/eth@19000000/mac-address <== this one
/proc/device-tree/ahb/eth@19000000/mtd-mac-address
/proc/device-tree/ahb/apb/wmac@18100000/mac-address
/proc/device-tree/ahb/apb/wmac@18100000/mtd-mac-address
```

There may be one, two or no paths giving the correct address. It may also happen that the `find` command returns nothing. Thanks.

adrianschmutzler commented:
Unfortunately, my description is slightly incorrect. The MAC addresses are stored binary, so you can't read with `cat`, but have to use the following code to read from the paths retrieved by `find`:
```
. /lib/functions/system.sh
get_mac_binary "$yourpath" 0
```
The include is obviously only required once. Sorry.
